File: ruby/021-02_ComeBackInOnePiece.rb
# 021 - Come Back in One Piece（★5）
# 強連結成分分解

class SCC
  attr_reader :size, :edges

  def initialize(n)
    @size = n
    @edges = Array.new(@size) { [] }
  end

  def add_edge(from, to) = edges[from] << to

  def add_edges(arr)
    arr.each { add_edge(_1, _2) }
  end

  def scc
    group_num, ids = scc_ids
    groups = Array.new(group_num) { [] }
    ids.each_with_index { groups[_1] << _2 }
    groups
  end

  private
  def scc_ids
    cur_ord = group_num = 0
    visited = []
    low = []
    ord = []
    ids = []

    dfs = -> from do
      low[from] = ord[from] = cur_ord
      cur_ord += 1
      visited << from
      edges[from].each do |to|
        if ord[to]
          low[from] = [low[from], ord[to]].min
        else
          dfs.(to)
          low[from] = [low[from], low[to]].min
        end
      end

      return unless low[from] == ord[from]

      until visited.empty?
        to = visited.pop
        ord[to] = size
        ids[to] = group_num
        break if to == from
      end
      group_num += 1
    end

    size.times do |i|
      dfs.(i) unless ord[i]
    end

    ids.map! { group_num - 1 - _1 }
    [group_num, ids]
  end
end

N, M = gets.split.map(&:to_i)

scc = SCC.new(N)

Array.new(M) { gets.split.map(&:to_i).map(&:pred) }.then { scc.add_edges(_1) }

ans = scc.scc.sum do |group|
  group.size * (group.size - 1) / 2
end

puts ans

File: kotlin/021-02_ComeBackInOnePiece.kt
/**
 * 021 - Come Back in One Piece（★5）
 * 強連結成分分解
 */

class SCC(val size: Int) {

  val edges = Array(size) { mutableListOf<Int>() }

  fun addEdge(from: Int, to: Int) {
    edges[from].add(to)
  }

  fun addEdges(pairs: List<Pair<Int, Int>>) {
    pairs.forEach { (from, to) -> addEdge(from, to) }
  }

  fun scc(): List<List<Int>> {
    val (groupNum, ids) = decompose()
    val groups = MutableList(groupNum) { mutableListOf<Int>() }
    ids.forEachIndexed { i, id -> groups[id].add(i) }
    return groups
  }

  private fun decompose(): Pair<Int, List<Int>> {
    var curOrd = 0
    var groupNum = 0
    val visited = mutableListOf<Int>()
    val low = IntArray(size) { -1 }
    val ord = IntArray(size) { -1 }
    val ids = IntArray(size)

    val dfs = DeepRecursiveFunction<Int, Unit> { from ->
      low[from] = curOrd
      ord[from] = curOrd
      curOrd++
      visited.add(from)
      edges[from].forEach { to ->
        if (ord[to] != -1) {
          low[from] = Math.min(low[from], ord[to])
        } else {
          callRecursive(to)
          low[from] = Math.min(low[from], low[to])
        }
      }

      if (low[from] != ord[from]) return@DeepRecursiveFunction

      while (visited.isNotEmpty()) {
        val to = visited.removeLast()
        ord[to] = size
        ids[to] = groupNum
        if (to == from) break
      }
      groupNum++
    }

    for (to in 0 until size) { if (ord[to] == -1) dfs(to) }

    return groupNum to ids.map { groupNum - it - 1 }
  }
}

fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val scc = SCC(N)
  repeat(M) {
    val (a, b) = readLine()!!.split(" ").map(String::toInt).map(Int::dec)
    scc.addEdge(a, b)
  }
  val ans = scc.scc().sumOf { it.size.toLong() * (it.size - 1) / 2 }
  println(ans)
}
